ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -893,6 +893,14 @@ private CompletableFuture<RaftClientReply> staleReadAsync(RaftClientRequest requ
     return processQueryFuture(stateMachine.queryStale(request.getMessage(), minIndex), request);
   }
 
+  private CompletableFuture<RaftClientReply> readOnlyAsync(RaftClientRequest request) {
+    LeaderState leaderState = role.getLeaderStateNonNull();
+    return leaderState.getReadIndex()
```

Comment:
Change it to 
```
    return role.getLeaderStateNonNull().getReadIndex()
```

Why if the server is a Follower?

Reply:
If the server is a follower, it should first send RPC to leader to obtain leader's `readindex`. Leader should also do the heartbeats to confirm leadership. After follower get the current `readindex`, it simply waits for statemachine to advance and query. I plan to implement follower linearizable read in next PR.